Best Way to Handle This
STEP 1 — Keep Them Separate

You should treat:

number_integer → numerical feature
number_categorical → categorical feature
STEP 2 — Apply Different Processing
Numerical Column

Use:

imputation
scaling
num_cols = ['number_integer']
Categorical Column

Use:

fill missing values
encoding

In [1]:
import pandas as pd 
import numpy as np 

In [2]:
df=pd.read_csv("titanic_01-handling-missing-data.csv")
df.head()

,Cabin,Ticket,number,Survived
0,NaN,A/5 21171,5,0.0
1,C85,PC 17599,3,1.0
2,NaN,STON/O2. 3101282,6,1.0
3,C123,113803,3,1.0
4,NaN,373450,A,0.0


In [3]:
df['number'].unique()

array(['5', '3', '6', 'A', '2', '1', '4', nan], dtype=object)

In [4]:
df.isnull().sum()

Cabin       688
Ticket        1
number        1
Survived      1
dtype: int64

In [5]:
df['number_integer']=pd.to_numeric(df['number'],downcast='integer' , errors='coerce')
df['number_categorical']=np.where(df['number_integer'].isnull() , df['number'] , np.nan)

In [6]:
df.head(10)

,Cabin,Ticket,number,Survived,number_integer,number_categorical
0,NaN,A/5 21171,5,0.0,5.0,NaN
1,C85,PC 17599,3,1.0,3.0,NaN
2,NaN,STON/O2. 3101282,6,1.0,6.0,NaN
3,C123,113803,3,1.0,3.0,NaN
4,NaN,373450,A,0.0,NaN,A
5,NaN,330877,2,0.0,2.0,NaN
6,E46,17463,2,0.0,2.0,NaN
7,NaN,349909,5,0.0,5.0,NaN
8,NaN,347742,1,1.0,1.0,NaN
9,NaN,237736,A,1.0,NaN,A


Smart Feature Engineering for Cabin

Extract first letter:

df['Cabin_Type'] = df['Cabin'].str[0]

In [7]:
df['cabin_num']=df["Cabin"].str.extract('([\d]+)')
df['cabin_cat']=df['Cabin'].str[0]
df.head(10)

<>:1: SyntaxWarning: invalid escape sequence '\d'
<>:1: SyntaxWarning: invalid escape sequence '\d'
C:\Users\LENOVO\AppData\Local\Temp\ipykernel_12636\1006854396.py:1: SyntaxWarning: invalid escape sequence '\d'
  df['cabin_num']=df["Cabin"].str.extract('([\d]+)')


,Cabin,Ticket,number,Survived,number_integer,number_categorical,cabin_num,cabin_cat
0,NaN,A/5 21171,5,0.0,5.0,NaN,NaN,NaN
1,C85,PC 17599,3,1.0,3.0,NaN,85,C
2,NaN,STON/O2. 3101282,6,1.0,6.0,NaN,NaN,NaN
3,C123,113803,3,1.0,3.0,NaN,123,C
4,NaN,373450,A,0.0,NaN,A,NaN,NaN
5,NaN,330877,2,0.0,2.0,NaN,NaN,NaN
6,E46,17463,2,0.0,2.0,NaN,46,E
7,NaN,349909,5,0.0,5.0,NaN,NaN,NaN
8,NaN,347742,1,1.0,1.0,NaN,NaN,NaN
9,NaN,237736,A,1.0,NaN,A,NaN,NaN


Smart Feature Engineering for Ticket

Ticket values:

A/5 21171
PC 17599
STON/O2. 3101282

You can extract:

Ticket Prefix
df['Ticket_Prefix'] = df['Ticket'].str.extract('([A-Za-z./]+)')

In [18]:
# extract ticket prefix
df['ticket_prefix'] = df['Ticket'].str.extract('([A-Za-z0-9./]+)')

# fill missing prefixes
df['ticket_prefix'] = df['ticket_prefix'].fillna('NoPrefix')

# extract last numeric part
df['ticket_number'] = df['Ticket'].str.extract('(\d+$)')

# convert to numeric
df['ticket_number'] = pd.to_numeric(df['ticket_number'])

<>:8: SyntaxWarning: invalid escape sequence '\d'
<>:8: SyntaxWarning: invalid escape sequence '\d'
C:\Users\LENOVO\AppData\Local\Temp\ipykernel_12636\3736328033.py:8: SyntaxWarning: invalid escape sequence '\d'
  df['ticket_number'] = df['Ticket'].str.extract('(\d+$)')


In [19]:
df.head(10)

,Cabin,Ticket,number,Survived,number_integer,number_categorical,cabin_num,cabin_cat,ticket_prefix,ticket_number
0,NaN,A/5 21171,5,0.0,5.0,NaN,NaN,NaN,A/5,21171.0
1,C85,PC 17599,3,1.0,3.0,NaN,85,C,PC,17599.0
2,NaN,STON/O2. 3101282,6,1.0,6.0,NaN,NaN,NaN,STON/O2.,3101282.0
3,C123,113803,3,1.0,3.0,NaN,123,C,113803,113803.0
4,NaN,373450,A,0.0,NaN,A,NaN,NaN,373450,373450.0
5,NaN,330877,2,0.0,2.0,NaN,NaN,NaN,330877,330877.0
6,E46,17463,2,0.0,2.0,NaN,46,E,17463,17463.0
7,NaN,349909,5,0.0,5.0,NaN,NaN,NaN,349909,349909.0
8,NaN,347742,1,1.0,1.0,NaN,NaN,NaN,347742,347742.0
9,NaN,237736,A,1.0,NaN,A,NaN,NaN,237736,237736.0
